src/azul/auth.py

```diff
@@ -12,6 +24,42 @@ def identity(self) -> str:
         """
         raise NotImplementedError
 
+    _cls_field = '_cls'
+
+    def to_json(self) -> JSON:
+        """
+        >>> @attr.s(auto_attribs=True)
+        ... class Foo(Authentication):
+        ...     foo: str
+        ...     def identity(self) -> str:
+        ...         # noinspection PyUnresolvedReferences
+        ...         return self.foo
+        >>> f = Foo('bar')
+        >>> f
+        Foo(foo='bar')
+        >>> f.to_json()
+        {'foo': 'bar', '_cls': 'Foo'}
+        >>> Authentication.from_json(f.to_json())
+        Foo(foo='bar')
+        """
+        assert self._cls_field not in dir(self), self
+        return {**attr.asdict(self), self._cls_field: type(self).__name__}
+
+    @classmethod
+    def from_json(cls, json: JSON) -> 'Authentication':
+        json = copy_json(json)
```

Comment:
It might be cheaper to just let `_cls` be a field in every instance, even though it's technically redundant.

Reply:
If we add `_cls` as an instance attribute, we will need to provide it whenever we instantiate an instance of `Auth`, including outside of the `from_json` method. We'd need to add factory classmethods for every subclass and replace all existing constructor calls with them. So I think it's simpler to stick with the current approach.